src/param.c

```diff
@@ -1558,19 +1558,33 @@ PARSE_FUNC(store_scat_grid)
 PARSE_FUNC(surf)
 {
 	double mre,mim;
-	if (Narg!=2 && Narg!=3) NargError(Narg,"2 or 3");
-	ScanDoubleError(argv[1],&hsub);
-	TestPositive(hsub,"height above surface");
-	if (strcmp(argv[2],"inf")==0) {
-		if (Narg>2) PrintErrorHelp("Additional arguments are not allowed for '-surf ... inf'");
-		msubInf=true;
-	}
-	else {
-		if (Narg!=3) NargError(Narg,"total 3 or second argument 'inf'");
-		ScanDoubleError(argv[2],&mre);
-		ScanDoubleError(argv[3],&mim);
-		msub = mre + I*mim;
-	}
+    if (Narg > 3 * (MAX_N_LAYERS + 1)) PrintErrorHelp("Exceeded the maximum number of layers");
+    if (Narg <= 1) NargError(Narg,">1");
+    ScanDoubleError(argv[1],&sub.hP);
+    TestPositive(sub.hP,"height above surface");
+    int arg_pos = 2;
+    sub.N = 1;
+    if (Narg > 3) {
+        int i;
+        for (i = 0; (arg_pos = 2 + 3 * i) < Narg - 1; ++i) {
+            ScanDoubleError(argv[arg_pos], &sub.h[i]);
+            ScanDoubleError(argv[arg_pos + 1], &mre);
+            ScanDoubleError(argv[arg_pos + 2], &mim);
+            sub.m[i] = mre + I * mim;
+        }
+        sub.N += i;
+    }
+    if (strcmp(argv[arg_pos],"inf")==0) {
+        if (Narg>arg_pos) PrintErrorHelp("Additional arguments are not allowed for '-surf ... inf'");
+        sub.mInf=true;
+    }
+    else {
+        if (Narg != arg_pos + 1) NargError(Narg, "total 3 or second argument 'inf'");
```

Comment:
These tests locate inf only in last 2-3 positions. If inf appears earlier, it is not detected. 

I propose to calculate sub.N from Narg. Then test all argv (except maybe first sub.N) for inf and process all cases of its misplacement. Then, finally, to scan all h and m.